### PR TITLE
Voice Calculator

### LINKED ISSUE #1014: Voice Calculator
# Aim
Build a calculator using voice command

# Details
I am building a calculator in which you can use voice commands for doing the calculation. The calculator is able to perform both Arithmetic as well as Scientific calculations. I might add a GUI but that's not promised.

### PR DESCRIPTION
**Pull request by Arnab11917676**
# Description
This is a voice calculator where you can use your voice to do calculations of arithmetic and trigonometric expression along with the history of the total calculation. To see the history just say HISTORY or to clear the history say CLEAR. This calculator is capable enough to calculate linear expressions also.

## Have you read the [Contributing Guidelines on Pull Requests](https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/master/CONTRIBUTING.md)?

- [x] Yes
- [ ] No

## Type of change

_Please delete options that are not relevant._

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines(Clean Code) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have created a helpful and easy to understand `README.md`
- [x] My documentation follows [`Template for README.md`](https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/master/Template%20for%20README.md)
- [x] My changes generate no new warnings
- [x] I have added tests/screenshots(if any) that prove my fix is effective or that my feature works.

## Project Metadata:

If there is no-file/nothing to fill the below fields with, then type: none

Example: If no requirements.txt needed/present then type in Requirments none => Requirments: none

Category:

- [x] Calculators
- [ ] AI/ML
- [ ] Scrappers
- [ ] Social_Media
- [ ] PDF
- [ ] Image_Processing
- [ ] Video_Processing
- [ ] Games
- [ ] Networking
- [x] OS_Utilities
- [ ] Automation
- [ ] Cryptography
- [ ] Fun
- [ ] Computer_Vision
- [ ] Others

Title: Voice Calculator

Folder: Voice Calculator

Requirements: requirements.txt

Script: voiceCalc.py

Arguments: none

Contributor: Arnab11917676

Description: A terminal-based voice calculator where you can do trigonometric, linear, and arithmetic calculations using voice.

### CODE REVIEW

**Comment by Arnab11917676** [comment]
@kaustubhgupta  have a look at it...

**Comment by Arnab11917676** [comment]
@kaustubhgupta  please share your thoughts...

File: Voice Calculator/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# Voice Calculator
+This is voice calculator capable enough to do difficult trigonometric probelms and simple arithmatic calculation and also Linear Expression.
+This will run in the terminal and all the instruction will be given while running it.
+
+
+## Setup instructions
+There are two ways to run it on your Linux,MAC or Windows
+
+- Run the executable(.exe) file.
```
Comment (kaustubhgupta) [inline Voice Calculator/README.md:9]: Is there any exe file here?

File: Voice Calculator/README.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+```
+```bash
+pip install speech_recognition
+```
+```bash
+pip install wolframalpha
+``` 
+```bash
+pip install tkinter
+```
```
Comment (kaustubhgupta) [inline Voice Calculator/README.md:26]: put the requirements in a separate file (requirements.txt)

File: Voice Calculator/README.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+|Voice Calculator| [Click Here](https://youtu.be/cOgujLzl9zg)| [Arnab Ray](https://github.com/Arnab11917676) | Speech Recognition , Wolframaplha
+
+## Output
+
+![image](https://user-images.githubusercontent.com/59610398/117536098-ccff8900-b016-11eb-809d-a63a3c23b39b.png)
+
+
+## Author(s)
+
+- Arnab Ray (Arnab11917676)
```
Comment (kaustubhgupta) [inline Voice Calculator/README.md:40]:
```suggestion
[Arnab Ray](https://github.com/Arnab11917676)
```

File: Voice Calculator/voiceCalc.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
+import os
+import pyttsx3
+import speech_recognition as sr
+import tkinter.messagebox as tmessage
+import wolframalpha
+
+
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:7]: why this extra spacing?

File: Voice Calculator/voiceCalc.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+from os.path import exists
+
+
+listener = sr.Recognizer()
+engine = pyttsx3.init()
+voices = engine.getProperty('voices')
+engine.setProperty('voice', voices[0].id)
+wolfprimeaplahe_app = '9RQ7T2-WKUWLR74WJ'
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:15]: why you have exposed the api key?

File: Voice Calculator/voiceCalc.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+welcomeInst()
+while True:
+
+    spech = _takeCommand().lower()
+
+    if 'calculate' in spech:
+        _calculate()
+
+
+
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:88]: extra space?

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]

File: Voice Calculator/voiceCalc.py (hunk 5 above)
Comment (Arnab11917676) [reply Voice Calculator/voiceCalc.py:15]: Because if I would not then the code would not have ran...

**Review by Arnab11917676 — COMMENTED** [review]
Please look at it
@kaustubhgupta

**Comment by Arnab11917676** [comment]
@kaustubhgupta  please look at it

File: Voice Calculator/README.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# Voice Calculator
+This is voice calculator capable enough to do difficult trigonometric probelms and simple arithmatic calculation and also Linear Expression.
+This will run in the terminal and all the instruction will be given while running it.
+
+
+## Setup instructions
+There are two ways to run it on your Linux,MAC or Windows
```
Comment (kaustubhgupta) [inline Voice Calculator/README.md:7]: Still two ways?

File: Voice Calculator/voiceCalc.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+import tkinter.messagebox as tmessage
+import wolframalpha
+
+from os.path import exists
+
+listener = sr.Recognizer()
+engine = pyttsx3.init()
+voices = engine.getProperty('voices')
+engine.setProperty('voice', voices[0].id)
+wolfprimeaplahe_app = 'USER_TOKEN'
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:13]: get this token from the user via input statements

File: Voice Calculator/README.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+- Run the program in your IDE or terminal/command line.
+
+## Requirements:
+Refer the requirments.txt file.
```
Comment (kaustubhgupta) [inline Voice Calculator/README.md:12]: remove these lines

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]
@Arnab11917676   Add the requirements.txt file and update the PR template with project metadata. Look at this PR for reference: https://github.com/avinashkranjan/Amazing-Python-Scripts/pull/1002

**Comment by Arnab11917676** [comment]
@kaustubhgupta  please look I tried to manage the issues. If I have messed once more please tell me.

File: Voice Calculator/README.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+This is voice calculator capable enough to do difficult trigonometric probelms and simple arithmatic calculation and also Linear Expression.
+This will run in the terminal and all the instruction will be given while running it.
+
+## Setup instructions
+
+There way to run it on your Linux,MAC or Windows
+
+- Run the program in your IDE or terminal/command line.
+
+<<<<<<< HEAD
```
Comment (kaustubhgupta) [inline Voice Calculator/README.md:12]: remove this line

File: Voice Calculator/README.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+Click on the **Click Here** to see the Voice Calculator Demo on YouTube.
+=======
```
Comment (kaustubhgupta) [inline Voice Calculator/README.md:14]: and this

File: Voice Calculator/README.md (hunk 3 above)
Comment (kaustubhgupta) [reply Voice Calculator/README.md:26]: @Arnab11917676 This suggest still left

File: Voice Calculator/voiceCalc.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+import tkinter.messagebox as tmessage
+import wolframalpha
+
+from os.path import exists
+
+listener = sr.Recognizer()
+engine = pyttsx3.init()
+voices = engine.getProperty('voices')
+engine.setProperty('voice', voices[0].id)
+wolfprimeaplahe_app = str(input('Enter the API Token'))
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:13]: All the inputs in python are by default string. Therefore no need to apply str func. 
```suggestion
wolfprimeaplahe_app = input('Enter the API Token')
```

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]

**Comment by Arnab11917676** [comment]
@kaustubhgupta  please look at the edits done...

**Comment by Arnab11917676** [comment]
@kaustubhgupta look at the changes made

File: Voice Calculator/README.md [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+| ---------------- | ------------------------------------------ | --------------------------------------------- | --------------------------------- |
+| Voice Calculator | [Click Here](https://youtu.be/cOgujLzl9zg) | [Arnab Ray](https://github.com/Arnab11917676) | Speech Recognition , Wolframalpha |
+
+## Output
+
+![image](https://user-images.githubusercontent.com/59610398/117536098-ccff8900-b016-11eb-809d-a63a3c23b39b.png)
+
+## Author(s)
+
+- Arnab Ray (https://github.com/Arnab11917676)
```
Comment (kaustubhgupta) [inline Voice Calculator/README.md:24]: still pending:
```suggestion
- [Arnab Ray](https://github.com/Arnab11917676)
```

File: Voice Calculator/voiceCalc.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    print('If you want calculate something please tell calcualte and then your expression')
+    audio('If you want calculate something please tell calcualte and then your expression')
+    print('For example CALCULATE 7 PLUS 8 or CALCULATE sin30 plus cot20')
+    audio('For example CALCULATE 7 PLUS 8 or CALCULATE sin30 plus cot20')
+
+
+def _takeCommand():
+    r = sr.Recognizer()
+    with sr.Microphone() as source:
+        print("Listning....")
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:33]:
```suggestion
        print("Listening....")
```

Also, use audio function to speak out this "Listening"

File: Voice Calculator/voiceCalc.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        r.pause_threshold = 2
+        r.energy_threshold = 3000
+        audio = r.listen(source)
+
+    try:
+        print("Recognizing...")
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:39]: After printing, use audio function to speak out this "Recognizing"

File: Voice Calculator/voiceCalc.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        query = r.recognize_google(audio, language='en-In')
+        print(query)
+
+    except Exception as e:
+        print(e)
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:44]: Don't print this, let's make the terminal clean

File: Voice Calculator/voiceCalc.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+        print("Did't understand you...\nPlease say again...")
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:45]:
```suggestion
        print("Didn't understand you...\nCan you repeat?")
```
Also, use the audio function to speak out this "Didn't understand you. Can you repeat?"

File: Voice Calculator/voiceCalc.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+    res = client.query(''.join(query))
+    answerr = next(res.results).text
+    space = '\n'
+    ourQuery = ''.join(query)
+    Question = 'Your Query was :- '
+    Answer = 'Your answer was :- '
+    finalAnswer = Question + str(ourQuery) + \
+        space + Answer + str(answerr) + space
+
+    if exists('maths.txt'):
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:64]:
```suggestion
    if exists('./Voice Calculator/maths.txt'):
```

File: Voice Calculator/voiceCalc.py [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+        with open('maths.txt', 'a', encoding='utf-8') as mth:
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:65]:
```suggestion
        with open('./Voice Calculator/maths.txt', 'a', encoding='utf-8') as mth:
```

File: Voice Calculator/voiceCalc.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+            mth.write(finalAnswer)
+            mth.close()
+    else:
+        history = open('maths.txt', 'w', encoding='utf-8')
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:69]:
```suggestion
        history = open('./Voice Calculator/maths.txt', 'w', encoding='utf-8')
```

File: Voice Calculator/voiceCalc.py [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+while True:
+
+    spech = _takeCommand().lower()
+
+    if 'calculate' in spech:
+        _calculate()
+
+    elif 'clear' in spech:
+
+        if exists('maths.txt'):
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:87]:
```suggestion
        if exists('./Voice Calculator/maths.txt'):
```

File: Voice Calculator/voiceCalc.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+            with open('maths.txt', 'r+') as file:
```
Comment (kaustubhgupta) [inline Voice Calculator/voiceCalc.py:88]:
```suggestion
            with open('./Voice Calculator/maths.txt', 'r+') as file:
```

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]
Do read **ALL** the **SUGGESTIONS** and make the changes:

File: Voice Calculator/voiceCalc.py (hunk 17 above)
Comment (Arnab11917676) [reply Voice Calculator/voiceCalc.py:45]: actually  Didn't is the correct grammatical form to represent the compressed form of [Did Not]

Comment (Arnab11917676) [reply Voice Calculator/voiceCalc.py:45]: @kaustubhgupta

**Comment by Arnab11917676** [comment]
@kaustubhgupta have a look on the suggested changes

**Review by kaustubhgupta — APPROVED** [review]

**Review by antrikshmisri — APPROVED** [review]

**Review by santushtisharma10 — APPROVED** [review]

**Comment by kaustubhgupta** [comment]
@Arnab11917676 who told you to make changes?

**Comment by kaustubhgupta** [comment]
I have put the approved tag that means the PR will be merged soon. Now rollback your last change

**Comment by Arnab11917676** [comment]
Sorry for the changes...I will refractor the changes back to what it was....

**Review by kaustubhgupta — APPROVED** [review]

**Comment by Arnab11917676** [comment]
@santushtisharma10  please check the PR

**Comment by kaustubhgupta** [comment]
> @santushtisharma10 please check the PR

Nothing to check here. The PR will be merged by PA soon. Kindly have patience.

**Comment by Arnab11917676** [comment]
@kaustubhgupta it's on hold for a long time...

**Review by avinashkranjan — APPROVED** [review]

**Pull request merged**